crates/amaru-consensus/src/chain_forward/client_protocol.rs

```diff
@@ -0,0 +1,249 @@
+use super::client_state::{find_headers_between, ClientOp, ClientState};
+use crate::consensus::store::ChainStore;
+use acto::{ActoCell, ActoInput, ActoRef, ActoRuntime, PanicOrAbort};
+use amaru_kernel::{to_cbor, Header};
+use pallas_network::{
+    facades::PeerServer,
+    miniprotocols::{
+        blockfetch,
+        chainsync::{self, ClientRequest, HeaderContent, Tip},
+        keepalive, txsubmission,
+    },
+};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ClientError {
+    #[error("client asked for headers before intersection was found")]
+    EarlyRequestNext,
+    #[error("no intersection found")]
+    NoIntersection,
+    #[error("client asked for intersection after it was already found")]
+    LateIntersection,
+    #[error("client terminated")]
+    ClientTerminated,
+    #[error("handler failure: {0}")]
+    HandlerFailure(PanicOrAbort),
+    #[error("chainsync error: {0}")]
+    ChainSync(chainsync::ServerError),
+    #[error("block fetch error: {0}")]
+    BlockFetch(blockfetch::ServerError),
+    #[error("tx submission error: {0}")]
+    TxSubmission(txsubmission::Error),
+    #[error("keep alive error: {0}")]
+    KeepAlive(keepalive::ServerError),
+}
+
+impl From<chainsync::ServerError> for ClientError {
+    fn from(value: chainsync::ServerError) -> Self {
+        ClientError::ChainSync(value)
+    }
+}
+
+impl From<blockfetch::ServerError> for ClientError {
+    fn from(value: blockfetch::ServerError) -> Self {
+        ClientError::BlockFetch(value)
+    }
+}
+
+impl From<txsubmission::Error> for ClientError {
+    fn from(value: txsubmission::Error) -> Self {
+        ClientError::TxSubmission(value)
+    }
+}
+
+impl From<keepalive::ServerError> for ClientError {
+    fn from(value: keepalive::ServerError) -> Self {
+        ClientError::KeepAlive(value)
+    }
+}
+
+pub async fn client_protocols(
+    mut cell: ActoCell<ChainSyncMsg, impl ActoRuntime, Result<(), ClientError>>,
+    server: PeerServer,
+    store: Arc<Mutex<dyn ChainStore<Header>>>,
```

Comment:
It might be worth thinking about introducing some read-only interface for the ChainStore eventually? Here, we'll have to lock and wait on the mutex every time clients wants to interact with the database -- also blocking the main thread from progressing. 

If anything, serving any of the chain-sync client should have a lower priority than the main job. So we need some kind of priority Mutex that ensures the main job is served _first_ when possible. Clients are secondary though since they don't need to write anything to the store in principle, we could have them work from some read-only structure and enable (1) more parallelism in the serving of the clients, and (2) avoid blocking the main job needlessly.

Reply:
Yes, I created #194 to track this.